### DISCUSSION TITLE
lookupUpdates error for Gemfile

### DISCUSSION THREAD

**Discussion by mvitz** [renovatebot/renovate#29820] [Request Help] [answered]
### What would you like help with?

I think I found a bug

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

Self-hosted GitLab, Renovate v37.415.0

### Please tell us more about your question or problem

Since Friday, our internal renovate job fails most of the time with a `lookupUpdates` error (see Logs).

I'm unfortunately unable to pin the error to a specific repository or dependency because these change from time to time. And I already suspected and tested en-/disabling caching between multiple runs, but this did not fix it either.

If more configuration settings are required for testing this, I'm happy to provide these.

Sorry that I'm unable to provide a reproducer or any more information. Hope somebody can help me.

### Logs (if relevant)

<details><summary>Logs</summary>

```
...
ERROR: lookupUpdates error (repository=XXX, packageFile=Gemfile)
       "currentDigest": undefined,
       "currentValue": "'~> 9.4'",
       "datasource": "rubygems",
       "packageName": "octicons_helper",
       "digestOneAndOnly": undefined,
       "followTag": null,
       "lockedVersion": "9.6.0",
       "pinDigests": false,
       "rollbackPrs": false,
       "isVulnerabilityAlert": undefined,
       "updatePinnedDependencies": true,
       "unconstrainedValue": false,
       "err": {
         "message": "Cannot read properties of null (reading 'compare')",
         "stack": "TypeError: Cannot read properties of null (reading 'compare')\n    at gt (/usr/local/renovate/node_modules/.pnpm/@renovatebot+ruby-semver@3.0.23/node_modules/@renovatebot/ruby-semver/dist/index.js:21:14)\n    at Object.isGreaterThan (/usr/local/renovate/lib/modules/versioning/ruby/index.ts:52:5)\n    at /usr/local/renovate/lib/workers/repository/process/lookup/filter.ts:41:18\n    at Array.filter (<anonymous>)\n    at filterVersions (/usr/local/renovate/lib/workers/repository/process/lookup/filter.ts:38:35)\n    at lookupUpdates (/usr/local/renovate/lib/workers/repository/process/lookup/index.ts:350:44)\n    at Function.wrap (/usr/local/renovate/lib/util/stats.ts:38:20)\n    at fetchDepUpdates (/usr/local/renovate/lib/workers/repository/process/fetch.ts:58:42)\n    at /usr/local/renovate/lib/workers/repository/process/fetch.ts:104:23\n    at /usr/local/renovate/node_modules/.pnpm/p-map@4.0.0/node_modules/p-map/index.js:57:22"
       }
...
DEBUG: Rubygems: error fetching rubygems data, falling back to versions-only result (repository=XXX)
       "err": {"type": "cache-invalid"}
...
```

</details>

**Comment by rarkins** [comment]
It seems like theres a bug in the underlying Ruby semver library but it's very hard to know how to fix it without having a reproduction or knowing which version triggered it.

**Reply by mvitz** [reply]
Is there anything I can do to provide additional information? I tried running it with TRACE logging enabled, but because we have many repositories, this produces a multiple GB log file.
Maybe preserving the cache to look into a file there or something?

**Reply by rarkins** [reply]
Work out which dependency causes it, then try to reproduce it in a test repo

**Comment by Goltergaul** [comment]
I'm seeing the same issue with our repositories. I'll try if i can find some additional infos

**Comment by ChristianCiach** [comment]
I updated from 38.25.1 to 38.40.0 and this seems to have fixed it.

**Reply by mvitz** [reply]
For us, the first three consecutive runs all succeeded, but just a few seconds ago the fourth run failed with the lookupUpdates error again.

**Comment by BjoernPetersen** [comment]
We're observing the same error (currently on Renovate v38.39.6) for projects with a `Gemfile` that's literally just this:

```
source "https://rubygems.org"

gem "fastlane"
```

[Corresponding Gemfile.lock](https://github.com/user-attachments/files/16658669/Gemfile.lock.txt) (had to add a .txt ending so GitHub accepts the file)

The error only appears sometimes, so it's hard to reproduce or pin down the root cause. The bug was introduced somewhere between Renovate v37.89.7 and v38.20.1 (we jumped between those versions due to a template change).

**Reply by mvitz** [reply]
Do you only check that single project or do you have multiple projects that are checked within a run?

**Reply by BjoernPetersen** [reply]
Multiple projects are checked within a run, 31 to be exact.

Also (just throwing it out there, might be irrelevant), for us, this happens in Flutter projects that have two Gemfiles, one at `android/Gemfile` and one at `ios/Gemfile`. We don't have any repos with a single Gemfile, so I can't tell whether that makes a difference.

**Reply by ChristianCiach** [reply]
We have hundreds of repos (within a single run), but only **one** of them is a Ruby project that contains a single `Gemfile`. This is enough to trigger this issue.

**Reply by mvitz** [reply]
That's an interesting observation. My best guess until now was that this would occur when multiple ruby projects were checked within one run. But your observation rules that one out.

**Reply by ChristianCiach** [reply]
I am sorry, I looked again. There is actually one other `Gemfile` hidden deep inside another repository that is scanned by Renovate (but in the same Gitlab group, so there is a chance that both repos are processed shortly after another). So please don't rule out your idea just yet :)

**Reply by ChristianCiach** [reply]
I cannot reproduce this locally. Do you guys use some kind of caching? We use both [`repositoryCache`](https://docs.renovatebot.com/self-hosted-configuration/#repositorycache) (which was once "experimental", but seems to be stable now) and [`redisUrl`](https://docs.renovatebot.com/self-hosted-configuration/#redisurl).

**Reply by BjoernPetersen** [reply]
We're using this exact template: https://gitlab.com/renovate-bot/renovate-runner, so we're using repository caching.

**Reply by mvitz** [reply]
I wasn't able to reproduce this locally once. The only thing I noticed was that, when running within the Docker Image, I wasn't able to execute renovate more than once because of some circumstances the Node.js version changed during the first run and wasn't able to start renovate after that.

When trying to reproduce this, the first thing I did was to disable the `repositoryCache`. But this didn't help. Afterwards, I tried to reproduce the error with a single repository, which didn't work either. Thereafter, I added one repository more. This didn't reproduce the error, either. I repeated that step and after some number of repositories the error occurred once again. I then attempted to only ignore that repo, but it didn't help either.

Since then, I gave up because I've no clue how to debug this any further.

**Reply by viceice** [reply]
> I wasn't able to reproduce this locally once. The only thing I noticed was that, when running within the Docker Image, I wasn't able to execute renovate more than once because of some circumstances the Node.js version changed during the first run and wasn't able to start renovate after that.

which version are you using? the latest docker images are using a separate node version to run renovate, so multiple runs should be possible.

> 
> When trying to reproduce this, the first thing I did was to disable the `repositoryCache`. But this didn't help. Afterwards, I tried to reproduce the error with a single repository, which didn't work either. Thereafter, I added one repository more. This didn't reproduce the error, either. I repeated that step and after some number of repositories the error occurred once again. I then attempted to only ignore that repo, but it didn't help either.
> 
> Since then, I gave up because I've no clue how to debug this any further.

**Comment by rybnico** [comment]
We also ran into this bug and this is what I found after some debugging/testing.
I nailed it down to the `vtrim` function - sometimes the string returned by this function doesn't contain the expected (trimmed) string, but some gibberish instead. I was able to catch the bug in the debugger. What I did was to compare the version returned by the cached `regEx` with an uncached one.

![image](https://github.com/user-attachments/assets/ec2f6b06-2ebe-465c-b2c9-4edf3cede7e7)

As you can see in the attached screenshot, the `version.replace` for `newVersion` returns some invalid characters/bytes, while the `version.replace` with the uncached `regEx` for `newVersionUncached` returns the expected string.

The problematic code can be found here: https://github.com/renovatebot/renovate/blob/bf2f3c3ca63db520e64bdce57f6cae7129857008/lib/util/regex.ts#L49

The cached `regEx` seems to be corrupted somehow.
Any idea why?

**Reply by ChristianCiach** [reply]
Since we cannot reproduce it, I wonder if maybe one of these URLs just return some corrupted contents sometimes:

```
DEBUG: http statistics (repository=local)
       "urls": {
         "https://rubygems.org/api/v1/gems/activesupport.json (GET,200)": 1,
         "https://rubygems.org/api/v1/gems/sinatra-cross_origin.json (GET,200)": 1,
         "https://rubygems.org/api/v1/gems/sinatra.json (GET,200)": 1,
         "https://rubygems.org/api/v1/versions/activesupport.json (GET,200)": 1,
         "https://rubygems.org/api/v1/versions/sinatra-cross_origin.json (GET,200)": 1,
         "https://rubygems.org/api/v1/versions/sinatra.json (GET,200)": 1,
         "https://rubygems.org/versions (GET,200)": 1
       },
```

I've queried these URLs many times and didn't see anything strange, but I am really out of ideas.

**EDIT** Sorry, please disregard that. I now see that in your example, `version` is still fine but `newVersion` gets corrupted by the regex replaces. This surely has nothing to do with the API responses.

**Reply by viceice** [reply]
we need to create a test case with normal regex and re2, so we can see if it's a cache bug or a re2 bug

**Reply by rybnico** [reply]
I had a go at creating a test case for this, but had no success. I suspect it has something to do with garbage collection/memory management.

**Reply by viceice** [reply, marked as answer]
I think this is now fixed

- https://github.com/uhop/node-re2/issues/214
- https://github.com/renovatebot/renovate/pull/31131

can you please verify?

**Reply by mvitz** [reply]
At least since updating to [38.58.2](https://github.com/renovatebot/renovate/releases/tag/38.58.2) which contains the mentioned fix, none of our pipelines failed with a lookupUpdates error. Therefore, I would assume that this is indeed fixed by the re2 update.
